controllers/storageautoscaler/verify.go

```diff
@@ -0,0 +1,137 @@
+	storageAutoScaler.Status.Phase = "Succeeded"
+	storageAutoScaler.Status.Error = nil
+	storageAutoScaler.Status.LastExpansion.CompletionTime = metav1.Now()
+	storageAutoScaler.Status.LastExpansion.StartOsdSize = expectedOsdSize
+	storageAutoScaler.Status.LastExpansion.StartOsdCount = expectedOsdCount
+	storageAutoScaler.Status.LastExpansion.StartStorageCapacity = expectedStorageCapacity
+
+	err := r.updateStatus(ctx, storageAutoScaler, originalStorageAutoScaler)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	r.Log.Info("scaling verified")
+
+	return reconcile.Result{}, nil
+}
+
+func (r *StorageAutoscalerReconciler) verifyVerticalScaling(ctx context.Context, storageAutoScaler *ocsv1.StorageAutoScaler, expectedOsdSize resource.Quantity) error {
+	osdCountQuery := "(ceph_osd_metadata * on (ceph_daemon, namespace, managedBy) group_right(device_class,hostname) (ceph_osd_stat_bytes))"
+	metrics, err := scrapeMetrics(ctx, r.OperatorNamespace, osdCountQuery, r.Log)
+	if err != nil {
+		r.Log.Error(err, "failed to scrape metrics")
+		return err
+	}
+
+	verified := false
+	for _, osd := range metrics {
+		deviceClass := osd.Metric["device_class"]
+		if string(deviceClass) == storageAutoScaler.Spec.DeviceClass {
+			verified = true
+			if expectedOsdSize.Cmp(resource.MustParse(osd.Value.String())) > 0 {
+				r.Log.Info("osd size not as expected", "expected", expectedOsdSize, "actual", osd.Value)
+				return errors.New("osd size not as expected, expected: " + expectedOsdSize.String() + " actual: " + osd.Value.String())
+			}
+			break
+		}
+	}
+
+	if !verified {
+		r.Log.Info("device class not found in metrics", "device class", storageAutoScaler.Spec.DeviceClass)
+		return errors.New("device class not found in metrics, device class: " + storageAutoScaler.Spec.DeviceClass)
+	}
+
+	return nil
+}
+
+func (r *StorageAutoscalerReconciler) verifyHorizontalScaling(ctx context.Context, storageAutoScaler *ocsv1.StorageAutoScaler, expectedOsdCount uint16) error {
+	osdCountQuery := "count by (device_class, namespace, managedBy) (ceph_osd_metadata)"
+	metrics, err := scrapeMetrics(ctx, r.OperatorNamespace, osdCountQuery, r.Log)
+	if err != nil {
+		r.Log.Error(err, "failed to scrape metrics")
+		return err
+	}
+
+	verified := false
+	for _, osd := range metrics {
+		deviceClass := osd.Metric["device_class"]
+		verified = true
+		if string(deviceClass) == storageAutoScaler.Spec.DeviceClass {
+			if float64(osd.Value) != float64(expectedOsdCount) {
+				r.Log.Info("osd count not as expected", "expected", expectedOsdCount, "actual", osd.Value)
+				return errors.New("osd count not as expected")
+			}
+			break
+		}
+	}
+
+	if !verified {
+		r.Log.Info("device class not found in metrics", "device class", storageAutoScaler.Spec.DeviceClass)
+		return errors.New("device class not found in metrics, device class: " + storageAutoScaler.Spec.DeviceClass)
+	}
+
+	return nil
+}
+
+func failIfTimeOut(timeoutSeconds int, reconcileStartTime metav1.Time) bool {
```

Comment:
"fail if timeout" suggest to me that this function would return an error, rather than a bool. Maybe "timeout has elapsed" or something similar would be a more indicative naming

```suggestion
func timeoutHasElapsed(timeoutSeconds int, reconcileStartTime metav1.Time) bool {
```